testing/e2e/infra/provision.go

```diff
@@ -0,0 +1,230 @@
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/keyvault/armkeyvault"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	// lenZones is the number of zones to provision
+	lenZones = 2
+	// lenPrivateZones is the number of private zones to provision
+	lenPrivateZones = 2
+)
+
+func (i *Infra) Provision(ctx context.Context) (ProvisionedInfra, *logger.LoggedError) {
+	lgr := logger.FromContext(ctx)
+	lgr.Info("provisioning infrastructure " + i.Name)
+	defer lgr.Info("finished provisioning infrastructure " + i.Name)
+
+	ret := ProvisionedInfra{
+		Name: i.Name,
+	}
+
+	var err error
+	ret.ResourceGroup, err = clients.NewResourceGroup(ctx, config.Flags.SubscriptionId, i.ResourceGroup, i.Location, clients.DeleteAfterOpt(2*time.Hour))
+	if err != nil {
+		return ProvisionedInfra{}, logger.Error(lgr, fmt.Errorf("creating resource group %s: %w", i.ResourceGroup, err))
+	}
+
+	// create resources
+	var resEg errgroup.Group
+	resEg.Go(func() error {
+		ret.ContainerRegistry, err = clients.NewAcr(ctx, config.Flags.SubscriptionId, i.ResourceGroup, "registry"+i.Suffix, i.Location)
+		if err != nil {
+			return logger.Error(lgr, fmt.Errorf("creating container registry: %w", err))
+		}
+		return nil
+	})
+
+	resEg.Go(func() error {
+		ret.Cluster, err = clients.NewAks(ctx, config.Flags.SubscriptionId, i.ResourceGroup, "cluster"+i.Suffix, i.Location, i.McOpts...)
+		if err != nil {
+			return logger.Error(lgr, fmt.Errorf("creating managed cluster: %w", err))
+		}
+		return nil
+	})
+
+	resEg.Go(func() error {
+		ret.KeyVault, err = clients.NewAkv(ctx, config.Flags.TenantId, config.Flags.SubscriptionId, i.ResourceGroup, "keyvault"+i.Suffix, i.Location)
+		if err != nil {
+			return logger.Error(lgr, fmt.Errorf("creating key vault: %w", err))
+		}
+
+		ret.Cert, err = ret.KeyVault.CreateCertificate(ctx, "cert"+i.Suffix, []string{"" + i.Suffix})
+		return nil
+	})
+
+	for idx := 0; idx < lenZones; idx++ {
+		func(idx int) {
+			resEg.Go(func() error {
+				zone, err := clients.NewZone(ctx, config.Flags.SubscriptionId, i.ResourceGroup, fmt.Sprintf("zone-%d-%s", idx, i.Suffix))
+				if err != nil {
+					return logger.Error(lgr, fmt.Errorf("creating zone: %w", err))
+				}
+				ret.Zones = append(ret.Zones, zone)
+				return nil
+			})
+		}(idx)
+	}
+	for idx := 0; idx < lenPrivateZones; idx++ {
+		func(idx int) {
+			resEg.Go(func() error {
+				privateZone, err := clients.NewPrivateZone(ctx, config.Flags.SubscriptionId, i.ResourceGroup, fmt.Sprintf("private-zone-%d-%s", idx, i.Suffix))
+				if err != nil {
+					return logger.Error(lgr, fmt.Errorf("creating private zone: %w", err))
+				}
+				ret.PrivateZones = append(ret.PrivateZones, privateZone)
+				return nil
+			})
+		}(idx)
+	}
+
+	if err := resEg.Wait(); err != nil {
+		return ProvisionedInfra{}, logger.Error(lgr, err)
+	}
+
+	// connect permissions
+	var permEg errgroup.Group
+
+	for _, pz := range ret.PrivateZones {
+		func(pz privateZone) {
```

Comment:
more a question than feedback, but why wrap this in a function of pz? Especially when the role assignment logic is already in a go func

Reply:
This is the common pattern for loop variable capture in Go.

Reply:
got it, if I remember correctly this is to avoid redeclaring the variable in the for loop, right?